configure.ac

```diff
@@ -29,6 +30,7 @@ NCCL_NET_OFI_DISTCHCK_CONFIGURE_FLAGS=
 # Checks for programs
 AC_PROG_CC
 m4_version_prereq([2.70], [], [AC_PROG_CC_STDC])
+AC_GNU_SOURCE
```

Comment:
I'm not actually sure why we need GNU_SOURCE defined, but I definitely don't want to make it the default.

Reply:
We definitely need it for asprintf, potentially others. 

I agree that we should avoid gnu extensions, but so long as they're already in the codebase, I feel it's more reasonable to commit to compiling with a consistent set of feature macros until we can get back to `_POSIX_C_SOURCE >= 200809L` compatibility.

Reply:
I feel strongly that this change is prudent because of potential ABI breakage when files are compiled with different feature test macros. 

I also went a step further and removed `config.h` for the same reason -- we have had bugs in the past few months where files did not include it, or didn't include it at the top, and this solves that for good -- no file is evaluated without bringing those macros into scope.